### ISSUE TITLE
Splitting out IHaskell.Display

### ISSUE THREAD

**Issue by abarbu** [gibiansky/IHaskell#1293]
Right now, libraries can't do anything smart about IHaskell, because they don't know they're running in IHaskell. They can't interact IHaskell without depending on the entire package. Other packages that face this issue usually provide a lightweight variant that only contains minimal definitions, enough to hook in, and then users can decide if they want to make use of that functionality or not.

What do you think about splitting out IHaskell.Display into its own lightweight package that doesn't depend on IHaskell or Jupyter? Anyone could make instances to display in the notebook independently of IHaskell. We could add `isRunningInIHaskell :: IO Bool` to let folks make smart runtime decisions.

With tools like jupytext that convert scripts back and forth, we can easily transition code in and out of IHaskell now. It's just the UX doesn't adapt properly.

This change would enable an ML library to decide to show graphs on screen with an image viewer or in a browser window when running in standalone mode. Or to show graphs in the notebook when running in IHaskell.

IHaskell.Display has very lightweight dependencies, none of which involve Jupyter. It needs DisplayData from ./ipython-kernel/src/IHaskell/IPython/Types.hs which is just `data DisplayData = DisplayData MimeType Text`. MimeType is just an enum.

It doesn't add any maintenance burden. If anything factoring out this code will make everything simpler.

What do you think? I'd be happy to do the legwork.

**Comment by vaibhavsagar** [comment]
I think we used to do this in the past: https://hackage.haskell.org/package/ihaskell-display, I don't know why we don't anymore but I'm assuming there was a good reason.

**Comment by vaibhavsagar** [comment]
Looking at the package I linked again, I think it is different from what you're suggesting because it pulled in `ihaskell` as a dependency instead of just `ipython-kernel`. Maybe we could even reorganise things so that `ipython-kernel` depends on `ipython-display` (or whatever we want to call it) and then we modify `IHaskell.Display` accordingly?

**Comment by abarbu** [comment]
Yeah, that's exactly what I was suggesting. Should be straightforward to reorganize everything. What you think?

**Comment by vaibhavsagar** [comment]
Hmm, honestly I don't have strong feelings about this change. I don't think it's necessary for the use case you stated, since you could instead provide an adapter package for notebook use that depends on both your ML library and IHaskell instead of always pulling in IHaskell.Display, and I also disagree that it reduces/doesn't change the maintenance burden, since we would have an extra package in the mix and that always complicates things. I think we would also have to inform e.g. `ihaskell-vega` if we did this, unless we re-exported the data type from IHaskell to make it transparent to consumers. At the same time, if you are convinced that this would make your life easier I'm definitely open to considering it, especially if we could make the change in a backwards-compatible way.

**Comment by habbler** [comment]
I think that the adapter package idea has a problem. The adapter package is dependent on all packages of both IHaskell and the user library. There can easily lead to a package version clash. It makes finding a consistent sent of package versions to build much more difficult (and leads to a large dependency tree).

Naturally the ihaskell package would re-export the ihaskell-display package. This should be a compatible change.

**Comment by vaibhavsagar** [comment]
We are very permissive about which versions of dependencies we use: https://github.com/gibiansky/IHaskell/blob/941e42d6fa58b37783da92061fb25befc16a52fa/ihaskell.cabal#L66-L97, so I don't think this is actually a problem in practice.

**Comment by jamesdbrock** [comment]
I think this is a really good idea.

**Comment by vaibhavsagar** [comment]
Could you say more about why you think this is a really good idea?

**Comment by jamesdbrock** [comment]
My understanding of what @abarbu proposed:

If we split the type `DisplayData` out into a separate package `ihaskell-displaydata`, then users can produce a `DisplayData` by just depending on `ihaskell-displaydata`.

And so user notebooks and `ihaskell` and `ipython-kernel` would all depend on `ihaskell-displaydata`.

Then, with `isRunningInIHaskell :: IO Bool`, functions which "display" things can voluntarily upgrade to `IHaskellDisplay` when they're in an IHaskell environment, without needing to pull in the entire `ihaskell` dependency even when they're not in an `IHaskell` environment.

I'm not sure how you would write `isRunningInIHaskell :: IO Bool` though.

---

Now that I've though about this, maybe you could get the same effect this way? Write all the functions which "display" things to return type `IO (MimeType, Text)`. Then you could write a function `displayIHaskell` to use when you're in an IHaskell notebook.

```haskell
displayIHaskell :: (MimeType, Text) -> Display
displayIHaskell = Display . pure . curry
```
 
https://github.com/IHaskell/IHaskell/blob/865df8c7fda0f5ee6af0a81e67690e1934e29ccf/src/IHaskell/Types.hs#L160-L161
